Solutions/Marain.Workflow.Deployment/Marain-PreDeploy.ps1

```diff
@@ -0,0 +1,25 @@
+﻿<#
+This is called during Marain.Instance infrastructure deployment prior to the Marain-ArmDeploy.ps
+script. It is our opportunity to perform initialization that needs to complete before any Azure
+resources are created.
+
+We create the Azure AD Application that the Operations Control function will use to authenticate incoming
+requests. (Currently, this application is used with Azure Easy Auth, but the service could also
+use it directly.)
+
+#>
+
+# Marain.Instance expects us to define just this one function.
+Function MarainDeployment([MarainServiceDeploymentContext] $ServiceDeploymentContext) {
+
+    $EngineApp = $ServiceDeploymentContext.DefineAzureAdAppForAppService("eng")
+    $MessageIngestionApp = $ServiceDeploymentContext.DefineAzureAdAppForAppService("mi")
+
+    $EngineControllerAppRoleId = "37a5c4e2-38e2-47de-8576-6b1ce7cc0ca2"
+    $EngineApp.EnsureAppRolesContain(
```

Comment:
Does this set up the permissions that this app has or does that need to be done seperately?

Reply:
Depends what you mean by "set up". This does not grant permissions. At this stage all we're doing is defining the kinds of permissions that exist for this particular app.

There's a necessary ordering here. We can't grant permissions until we know all the relevant ids: the service principal IDs of the services being granted permissions, and the ID of the enterprise app associated with the target app that defines the roles; this means we have to run the ARM deployment before we can assign permissions, because the service principals don't even exist until after that's been done. And we can't run the ARM deployment until the apps used for authenticating incoming requests exist.

So we have to define the apps used for authentication as the first step. And it makes sense to define all the app roles at the same point as defining the app, I believe. (Although in principle we could do that later. The only strict prerequisite for ARM deployment is that we know the App IDs.)

Reply:
Right yes, of course.

Reply:
Ah wait, sorry, I meant for the role itself, as in, does a role by default have access to all endpoints?

Reply:
An app role is really just a named thing that principals can be assigned to. A role doesn't do anything by itself.

If a principal (either a user or a service principal) is assigned to at least one app role, that means it will be possible to get an access token for that app, meaning that it will be possible to get past the barrier put up by Easy Auth (or an equivalent barrier where the app itself enforces a rule saying that incoming requests must have an access token for the app). Beyond that, it's entirely up to the application to decide what role membership actually implies.

So membership of a role implies access of some kind to the app that defined the role. (But remember, role membership is handled elsewhere entirely. At this point, we're just making sure that the roles have been defined. Role definitions live in the Application Manifest. Role membership assignments happen in the "enterprise app" (as the AAD portal calls it; technically it's actually a service principal, although that's horribly confusing) corresponding to the app. The reason for that is to support multi-tenant apps: when an app is used in multiple tenants, there is a single definition of what roles exist shared across all tenants, but each tenant the app is used in defines its own tenant-specific role memberships.)

To answer your question annoyingly literally, no, a role does not have access to anything, because in AAD's model, only principals have access to things, and a role is not a principle. A principle that is assigned to one or more of an app's roles will have access to that app. But the role itself is not a thing that can have access to anything. A role is really a name for a kind of permission a principal might be granted.